secure/storage/src/crypto_storage.rs

```diff
@@ -12,6 +13,7 @@ use serde::{Deserialize, Serialize};
 /// keys securely. The API offered here is inspired by the 'Transit Secret Engine' provided by
 /// Vault: a production-ready storage engine for cloud infrastructures (e.g., see
 /// https://www.vaultproject.io/docs/secrets/transit/index.html).
+#[enum_dispatch]
```

Comment:
Somewhat orthogonal question, but, have we thought about the security implications of adding the enum_dispatch crate to our dependencies? Secure storage is part of the TCB, and in general we're trying to reduce our dependencies (and avoid adding new ones). Part of me wonders what kinds of rust magic (e.g., unsafe code), the enum_dispatch crate uses to achieve its amazing performance boosts 😄

Reply:
So, I have audited the crate in question (and it's a proc-macro crate, so it really can't do anything but inject rust code) and it does not generate unsafe code. Whether it contains a use of unsafe is irrelevant, but I nonetheless checked it is not the case for enum_dispatch itself. 

> what kinds of rust magic

It generates the code that I replace, and adds `TryInto<X> for Storage` for every variant that `Storage` closes over. 

The performance gain brought by enum_dispatch is that brought by an enum closing over an open trait hierarchy, whether that enum is written manually or not. Hence that performance gain exists between the dynamic dispatch that existed pre-#4386, and the enum that exists in #4386. That change in performance profile is well understood (dynamic dispatch is costly).

The present PR does not change performance, it removes manual boilerplate and generates the same solution it with a macro.

The trade-off here is not simply whether we take a dependency or not, but whether the risk induced by generating this code using a macro is smaller or greater than letting such boilerplate hinder the maintenability of our code base.

If we wanted, we could also rewrite the boilerplate-generation ourselves (enum_dispatch was written a while ago and could use conveniences brought on by the latest syn/quote practices — also @matbd suggested he could use plain macros to do something similar).

Reply:
Awesome, thanks for looking into it and for the thorough explanation @huitseeker! This now makes sense to me.